File: src/behavioral/observer2.ts
enum Event {
  NEW_ITEM,
  SALE,
}

interface EventListener {
  update(eventType: Event): void;
}

class NotificationService {
  private readonly customers: Map<Event, EventListener[]> = new Map();

  constructor() {
    this.customers.set(Event.NEW_ITEM, []);
    this.customers.set(Event.SALE, []);
  }

  subscribe(eventType: Event, listener: EventListener): void {
    this.customers.get(eventType)?.push(listener);
  }

  unsubscribe(eventType: Event, listener: EventListener): void {
    // TypeScript arrays don't have a direct .remove() method.
    // We use .indexOf() and .splice() to remove the listener.
    const listeners = this.customers.get(eventType);
    if (listeners) {
      const index = listeners.indexOf(listener);
      if (index !== -1) {
        listeners.splice(index, 1);
      }
    }
  }

  notify(eventType: Event): void {
    this.customers?.get(eventType)?.forEach((listener) => {
      console.log({ eventType });
      console.log({ listener });
      return listener?.update(eventType);
    });
  }
}

class Store {
  private notificationService: NotificationService;

  constructor() {
    this.notificationService = new NotificationService();
  }

  getNotificationService(): NotificationService {
    return this.notificationService;
  }
}

class EmailMsgListener implements EventListener {
  private email: string;

  constructor(email: string) {
    this.email = email;
  }

  update(): void {
    // Actually send the email
    console.log("email sent");
  }
}

class MobileAppListener implements EventListener {
  private username: string;

  constructor(username: string) {
    this.username = username;
  }

  update(): void {
    // Actually send the push notification
    console.log("push notification sent");
  }
}

function notifyClients(): void {
  const store = new Store();

  store.getNotificationService().subscribe(Event.NEW_ITEM, new EmailMsgListener("geekific@like.com"));
  store.getNotificationService().subscribe(Event.SALE, new EmailMsgListener("geekific@subs.com"));
  store.getNotificationService().subscribe(Event.NEW_ITEM, new MobileAppListener("GeekificLnS"));

  store.getNotificationService().notify(Event.NEW_ITEM);
  store.getNotificationService().notify(Event.SALE);
}

notifyClients();

File: src/behavioral/observer2.java
public enum Event {
    NEW_ITEM, SALE
}

public interface EventListener {
    void update(Event eventType);
}

public class NotificationService {
    private final Map<Event, List<EventListener>> customers;

    public NotificationService() {
        customers = new HashMap<>();
        Arrays.stream(Event.values()).forEach(event ->
            customers.put(event, new ArrayList<>())
        );
    }

    public void subscribe(Event eventType, EventListener listener) {
        customers.get(eventType).add(listener);
    }

    public void unsubscribe(Event eventType, EventListener listener) {
        customers.get(eventType).remove(listener);
    }

    public void notify(Event eventType) {
        customers.get(eventType).forEach(listener ->
            listener.update(eventType)
        );
    }
}

public interface EventListener {
    void update();
}

public class EmailMsgListener implements EventListener {
    private final String email;

    public EmailMsgListener(String email) {
        this.email = email;
    }

    @Override
    public void update() {
        // Actually send the mail
    }
}

public class MobileAppListener implements EventListener {
    private final String username;

    public MobileAppListener(String username) {
        this.username = username;
    }

    @Override
    public void update() {
        // Actually send the push notification
    }
}

public static void main(String[] args) {
    Store store = new Store();
    store.getNotificationService().subscribe(
        new EmailMsgListener("geekific@like.com")
    );
    store.getNotificationService().subscribe(
        new EmailMsgListener("geekific@subs.com")
    );
    store.getNotificationService().subscribe(
        new MobileAppListener("GeekificLnS")
    );
}